### PR TITLE
Repartition OP

### LINKED ISSUE #1074: Update Repartition/Partition operator

### PR DESCRIPTION
**Pull request by lambda7xx**
**Description of changes:**
- update the Repartition OP

**Related Issues:**

Linked Issues:
- Issue #1074 

Issues closed by this PR:
- Closes #1074

- - -
This change is [Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1119)

### CODE REVIEW

**Review by wmdi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1119)* status: 1 of 5 files reviewed, 3 unresolved discussions (waiting on @lambda7xx, @lockshaw, and @reyna-abhyankar)

___
*[`lib/runtime/src/ops/partition.cc` line 72 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1119#-NfNSw9uCevi_clHxqOe:-NfNSw9uCevi_clHxqOf:b-9wroup) ([raw file](https://github.com/flexflow/FlexFlow/blob/06c96ad8d75ab27e529c332898232191083ad4cd/lib/runtime/src/ops/partition.cc#L72)):*
> ```cpp
>   OpTaskBinding binding;
> 
>   binding.bind_arg(HANDLE, ff_handle());
> ```

Where is `HANDLE`?
___
*[`lib/runtime/src/ops/partition.cc` line 89 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1119#-NfNUAuh-PKCz1JeXTRt:-NfNUAuh-PKCz1JeXTRu:b45h) ([raw file](https://github.com/flexflow/FlexFlow/blob/06c96ad8d75ab27e529c332898232191083ad4cd/lib/runtime/src/ops/partition.cc#L89)):*
> ```cpp
> }
> 
> OpTaskInvocation backward(CastAttrs const &attrs) {
> ```





_Suggestion:_
```cpp
OpTaskInvocation backward(RepartitionAttrs const &attrs) {
```
___
*[`lib/runtime/src/ops/partition.cc` line 92 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1119#-NfNXjvVCuP1Ax1F_7Af:-NfNXjvW6vJtojGUyG3V:b-koewhj) ([raw file](https://github.com/flexflow/FlexFlow/blob/06c96ad8d75ab27e529c332898232191083ad4cd/lib/runtime/src/ops/partition.cc#L92)):*
> ```cpp
>   OpTaskBinding binding = infer_bwd_binding(forward(attrs).binding);
> 
>   PerDeviceFFHandle handle = acc.get_argument<PerDeviceFFHandle>(HANDLE);
> ```

`handle` is unused

File: lib/runtime/src/ops/partition.cc [hunk 1]
```diff
@@ -62,295 +66,165 @@ RepartitionParams Repartition::get_params() const {
...
-                                                    0 /*projection id*/,
-                                                    WRITE_ONLY,
-                                                    EXCLUSIVE,
-                                                    outputs[0]->region));
-  launcher.add_field(1, FID_DATA);
-  runtime->execute_index_space(ctx, launcher);
+OpTaskInvocation backward(CastAttrs const &attrs) {
+  OpTaskBinding binding = infer_bwd_binding(forward(attrs).binding);
+
+  PerDeviceFFHandle handle = acc.get_argument<PerDeviceFFHandle>(HANDLE);
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/partition.cc:92]:
```suggestion

```

File: lib/runtime/src/ops/partition.cc [hunk 2]
```diff
@@ -40,6 +42,8 @@ using Legion::TaskLauncher;
 
 using namespace FlexFlow::Kernels::Repartition;
 
+enum Slots { INPUT, OUTPUT, ATTRS, PROFILING };
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/partition.cc:45]:
```suggestion
enum Slots { INPUT, OUTPUT, ATTRS, PROFILING, HANDLE };
```

File: lib/runtime/src/ops/partition.cc [hunk 3]
```diff
@@ -62,295 +66,165 @@ RepartitionParams Repartition::get_params() const {
...
+      infer_bwd_signature(get_op_signature(REPARTITION_FWD_TASK_ID));
 
-namespace std {
-size_t hash<FlexFlow::RepartitionParams>::operator()(
-    FlexFlow::RepartitionParams const &params) const {
-  size_t key = 0;
-  hash_combine(key, params.repartition_legion_dim);
-  hash_combine(key, params.repartition_degree);
-  return key;
+  registar_task(REPARTITION_BWD_TASK_ID, "Repartition Bwd", bwd, backward_task);
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/partition.cc:227]:
```suggestion
  register_task(REPARTITION_BWD_TASK_ID, "Repartition Bwd", bwd, backward_task);
```

**Review by reyna-abhyankar — CHANGES_REQUESTED** [review]

File: lib/runtime/src/ops/partition.cc (hunk 3 above)
Comment (lambda7xx) [reply lib/runtime/src/ops/partition.cc:227]: pushed

**Review by reyna-abhyankar — APPROVED** [review]

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/partition.cc` line 45 at r2](https://reviewable.io/reviews/flexflow/FlexFlow/1119#gh-1341912783:-NgzLuI-DDUmNdl471xw:baptkx5) ([raw file](https://github.com/flexflow/FlexFlow/blob/986af8faa8549de36e2e5da7937771447d2f7bc9/lib/runtime/src/ops/partition.cc#L45)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

```suggestion
enum Slots { INPUT, OUTPUT, ATTRS, PROFILING, HANDLE };
```
</blockquote></details>

we should add `PER_DEVICE_STATE` for forward and backward

**Review by lambda7xx — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1119)* status: 1 of 5 files reviewed, 6 unresolved discussions (waiting on @reyna-abhyankar and @wmdi)

___
*[`lib/runtime/src/ops/partition.cc` line 72 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1119#-NfNSw9uCevi_clHxqOe:-NgzLQTs31b2tcGU4VsT:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/06c96ad8d75ab27e529c332898232191083ad4cd/lib/runtime/src/ops/partition.cc#L72)):*
<details><summary><i>Previously, wmdi (Mengdi Wu) wrote…</i></summary><blockquote>

Where is `HANDLE`?
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/partition.cc` line 89 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1119#-NfNUAuh-PKCz1JeXTRt:-NgzLSIcDMCEQrSLAMvc:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/06c96ad8d75ab27e529c332898232191083ad4cd/lib/runtime/src/ops/partition.cc#L89)):*
> ```cpp
> }
> 
> OpTaskInvocation backward(CastAttrs const &attrs) {
> ```

Done.
___
*[`lib/runtime/src/ops/partition.cc` line 92 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1119#-NfNXjvVCuP1Ax1F_7Af:-NgzLeMwEl90w2f3_X8g:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/06c96ad8d75ab27e529c332898232191083ad4cd/lib/runtime/src/ops/partition.cc#L92)):*
<details><summary><i>Previously, wmdi (Mengdi Wu) wrote…</i></summary><blockquote>

`handle` is unused
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/partition.cc` line 92 at r2](https://reviewable.io/reviews/flexflow/FlexFlow/1119#gh-1341912722:-NgzLtZG3upDhP3YGxFX:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/986af8faa8549de36e2e5da7937771447d2f7bc9/lib/runtime/src/ops/partition.cc#L92)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

```suggestion

```
</blockquote></details>

Done.

**Review by wmdi — APPROVED** [review]
> Reviewed 2 of 5 files at r1, 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1119)* status: all files reviewed, 3 unresolved discussions (waiting on @reyna-abhyankar)

**Pull request merged**